WPB-17086: fix makefile to consider latest only if it exists in case … (#53)
wireapp/wire-docs: src/how-to/install/helm-prod.rst

.. code:: shell

   cd wire-server-deploy/
   wget -r -l 10 https://helm.internal.vpc/charts/
   mv helm.internal.vpc/charts/ ./wire
   rm $(find wire/ -name index*)

where 'internal.vpc' needs to be replaced with the domain you're using in your offline environment.

Preparing to install charts from wire with Helm 2
--------------------------------------------------
If your environment is online, and you wish to use helm2, you need to add the remote wire helm repository, to download wire charts.

To enable the wire charts helm repository:

.. code:: shell

   helm repo add wire https://s3-eu-west-1.amazonaws.com/public.wire.com/charts

(You can see available helm charts by running ``helm search wire/``. To see
new versions as time passes, you may need to run ``helm repo update``)

Great! Now you can start installing.

.. note::

    all commands below can also take an extra ``--namespace <your-namespace>`` if you don't want to install into the default kubernetes namespace.

> There is a shell script for doing the following procedure with helm 2, for reference. examine wire-server-deploy/bin/prod-setup.sh .

Reviewer: ```suggestion
There is a shell script for doing the following procedure with helm 2. For reference, examine `prod-setup.sh <https://github.com/wireapp/wire-server-deploy/blob/develop/bin/prod-setup.sh>`__.
```
Not sure about the syntax, though.

Watching changes as they happen
-------------------------------

Open a terminal and run:

.. code:: shell

   kubectl get pods -w

This will block your terminal and show some things happening as you proceed through this guide. Keep this terminal open and open a second terminal.

> How to install charts that provide access to external databases
> ---------------------------------------------------------------
> --------------------------------------------------------------------------------
> Before you can deploy the helm charts that tell wire where external services

Reviewer: ```suggestion
How to install charts that provide access to external databases
---------------------------------------------------------------

Before you can deploy the helm charts that tell wire where external services
```

   wire_server_deploy_commit:  01e0f261ca8163e63860f8b2af6d4ae329a32c14
   wire_server_deploy_link:    https://github.com/wireapp/wire-server-deploy/releases/tag/chart/wire-server-0.24.9

Note you'll need ``kubectl``, ``git`` and ``helm`` installed

It will output the running docker image; the corresponding wire-server
commit hash (and link) and the wire-server helm chart version which is
running. This will be helpful for any support requests.

Helm install / upgrade failed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Usually, you want to run::

    kubectl get pods --all-namespaces

And look for any pods that are not ``Running``. Then you can::

    kubectl --namespace <namespace> logs <name-of-pod>

and/or::

    kubectl --namespace <namespace> describe <name-of-pod>

to know more.

As long as nobody is using your cluster yet, you can safely delete and re-create a specific helm release (list releases with ``helm list --all``). Example delete the ``wire-server`` helm release:

.. code:: shell

    helm delete --purge wire-server

It doesn't work, but my problem isn't listed here. Help!
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Feel free to open a github issue or pull request `here <https://github.com/wireapp/wire-docs>`_ and we'll try to improve the documentation.

Reviewer: I'm undecided whether we want to point people to the docs issues or wire-server-deploy repo. Any thoughts? If people have problems deploying things, for which the code foundation resided in wire-server-deploy they might look for similar issues there and not here, wont they?

   Code in this repository should be considered *beta*. As of 2019, we do not (yet)
   run our production infrastructure on kubernetes (but plan to do so soon).

Introduction
-----------------

> The following will install a version of all the wire-server components. This setup is not recommended in production but will get you started.

Reviewer 2: what makes it not recommended in production?

Prerequisites
--------------------------------

You need to have access to a kubernetes cluster, and the ``helm`` local binary on your PATH.
Your kubernetes cluster needs to have internal dns services, so that wire-server can find it's databases.
You need to have docker on the machine you are using to perform this installation with, or a secure data path to a machine that runs docker. You will be using docker to generate security credentials for your wire installation.

* If you don't have a kubernetes cluster, you have two options:

  * You can get access to a managed kubernetes cluster with the cloud provider of your choice.
  * You can install one if you have ssh access to a virtual machine, see :ref:`ansible-kubernetes`

* If you don't have ``helm`` yet, see `Installing helm <https://helm.sh/docs/using_helm/#installing-helm>`__.

Type ``helm version``, you should, if everything is configured correctly, see a result like this:

::

    Client: &version.Version{SemVer:"v2.13.1", GitCommit:"618447cbf203d147601b4b9bd7f8c37a5d39fbb4", GitTreeState:"clean"}
    Server: &version.Version{SemVer:"v2.13.1", GitCommit:"618447cbf203d147601b4b9bd7f8c37a5d39fbb4", GitTreeState:"clean"}

> In case ``kubectl version`` shows both Client and Server versions, but ``helm version`` does not show a Server version, you may need to run ``helm init``. The exact version (assuming `v2.X.X` - at the time of writing v3 is not yet supported) matters less as long as both Client and Server versions match (or are very close).

Reviewer 2: also helm 3 does not have init to add to the confusion

Great, now try the installation:

.. code:: shell

   helm install wire-server wire/wire-server -f my-wire-server/values.yaml -f my-wire-server/secrets.yaml --wait

How to direct traffic to your cluster
------------------------------------------

There are a few options available. The easiest option is to use an ingress with a node port, as this works everywhere and doesn't need a special setup.

.. code:: shell

   # (assuming you're in the root directory of wire-server-deploy)
   mkdir -p nginx-ingress-services && cd nginx-ingress-services
   cp ../values/nginx-ingress-services/demo-secrets.example.yaml secrets.yaml
   cp ../values/nginx-ingress-services/demo-values.example.yaml values.yaml

You should now have the following directory structure:

::

  .
  ├── nginx-ingress-services
  │   ├── secrets.yaml
  │   └── values.yaml
  └── my-wire-server
      ├── secrets.yaml
      └── values.yaml

Inside the ``nginx-ingress-services`` directory, open ``values.yaml`` and replace ``example.com`` with a domain of your choosing. You can try using ``sed -i 's/example.com/<your-domain>/g' values.yaml``.

> Next, open ``secrets.yaml`` and add a TLS wildcard certificate and private key matching your domain. For ``example.com``, you need a certficate for ``*.example.com``. The easiest and cheapest options are:
> 
> 1. use `Let's Encrypt <https://letsencrypt.org/getting-started/>`__

Reviewer: ```suggestion
Next, open ``secrets.yaml`` and add a TLS wildcard certificate and private key matching your domain. For ``example.com``, you need a certficate for ``*.example.com``. The easiest and cheapest option is `Let's Encrypt <https://letsencrypt.org/getting-started/>`__
```
this one item in the list looks a bit lonely ;)
